### PR TITLE
NN-2892-cell-history

### PR DESCRIPTION
**Pull request by JamieDraperUK**
- New API to expose history at a cell location

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/CellSwapResult.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@ApiModel(description = "Cell swap result")
+@Builder(toBuilder = true)
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class CellSwapResult {
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/CellSwapResult.java:15]: Tidy up after a recent PR

File: src/main/java/uk/gov/justice/hmpps/prison/web/config/VersionOutputter.java [hunk 2]
```diff
@@ -10,7 +10,7 @@
 import org.springframework.context.annotation.Configuration;
 import org.springframework.context.event.EventListener;
 
-@Slf4j
+/*@Slf4j
 @Configuration
```
Comment (andrewrlee) [inline src/main/java/uk/gov/justice/hmpps/prison/web/config/VersionOutputter.java:14]: is this supposed to be commented out?

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/CellResource.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+@RequestMapping("${api.base.path}/cell")
+@Validated
+public interface CellResource {
+    @GetMapping("/{locationId}/history")
+    @ApiResponses(value = {
+            @ApiResponse(code = 400, message = "Invalid request.", response = ErrorResponse.class),
+            @ApiResponse(code = 404, message = "Requested resource not found.", response = ErrorResponse.class),
+            @ApiResponse(code = 500, message = "Unrecoverable error occurred whilst processing request.", response = ErrorResponse.class)})
+    List<BedAssignment> getBedAssignmentsHistory(
+            @ApiParam(value = "The offender booking linked to the court hearings.", required = true) @PathVariable("locationId") Long locationId,
```
Comment (matt-crowson-moj) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/CellResource.java:29]: court hearings?

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/CellResource.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+            @ApiParam(value = "From date", example = "2020-03-24", required = true) @RequestParam(value = "fromDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate fromDate,
+            @ApiParam(value = "To date", example = "2020-03-24", required = true) @RequestParam(value = "toDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate toDate);
```
Comment (matt-crowson-moj) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/CellResource.java:31]: Just my opinion but perhaps good to have different dates for the from and to. Meh.

File: src/main/java/uk/gov/justice/hmpps/prison/web/config/VersionOutputter.java (hunk 2 above)
Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/hmpps/prison/web/config/VersionOutputter.java:14]: No 😨

File: src/main/java/uk/gov/justice/hmpps/prison/service/BedAssignmentHistoryService.java [hunk 5]
```diff
@@ -51,22 +52,37 @@ public void add(final Long bookingId, final Long livingUnitId, final String reas
...
+        final var results = bedAssignmentsHistory.getContent()
+                .stream()
+                .map(this::transform)
+                .collect(Collectors.toList());
+
         return new PageImpl<>(results, pageRequest, bedAssignmentsHistory.getTotalElements());
     }
 
+    public List<BedAssignment> getBedAssignmentsHistory(final long livingUnitId, final LocalDate from, final LocalDate to) {
+        if (!locationRepository.existsById(livingUnitId)) throw new EntityNotFoundException(String.format("Cell %s not found", livingUnitId));
```
Comment (matt-crowson-moj) [inline src/main/java/uk/gov/justice/hmpps/prison/service/BedAssignmentHistoryService.java:64]: Perhaps check the dates are valid/correct way round?

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

File: src/main/java/uk/gov/justice/hmpps/prison/service/BedAssignmentHistoryService.java (hunk 5 above)
Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/hmpps/prison/service/BedAssignmentHistoryService.java:65]: Not sure we need to check the format as it's dealt with at the controller level. However, I agree that the date order should be validated.

**Comment by JamieDraperUK** [comment]
🍰

File: src/main/java/uk/gov/justice/hmpps/prison/service/BedAssignmentHistoryService.java (hunk 5 above)
Comment (matt-crowson-moj) [reply src/main/java/uk/gov/justice/hmpps/prison/service/BedAssignmentHistoryService.java:65]: Not sure on the logic here :

```
if (from.isAfter(to) && !from.isEqual(to)) {
  fail
}

```
You can just:

```
if (from.isAfter(to))  {
  fail
}
```

**Review by matt-crowson-moj — DISMISSED** [review]
LGTM

File: src/main/java/uk/gov/justice/hmpps/prison/service/BedAssignmentHistoryService.java (hunk 5 above)
Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/hmpps/prison/service/BedAssignmentHistoryService.java:65]: Yeah you're right.

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

**Pull request merged**